### ISSUE TITLE
Concurrent modification exception when merging segments

### ISSUE THREAD

**Issue by ptomli** [AxonIQ/axon-server-se#135]
I've run into a scenario where the server and framework aren't completely in agreement of the state of segments being processed. Or perhaps the framework just needs to be more aware of token store data being updated by other application instances.

This is repeatable in an application that's configured for 2 threads per instance, with a JPA token store, running against Server SE 4.3.3, with framework version 4.3.3.

When an application instance dies, and therefore does not properly update the token store, the token store still has old/invalid data saying the dead instance is processing X segments. A later request to merge a processor can result in the below exception. The age of the dead instance seems to somehow have an effect, but I can't figure out how.

A workaround appears to be to pause & restart the processor before merging, if the processor is reporting "Not all segments claimed". This results in the token store being updated and no error is encountered.

Sorry I've not been able to find the root cause, though I imagine someone with a more complete mental image of the interactions between server, application and token store may have better luck.

This was the subject of a question on StackOverflow  https://stackoverflow.com/questions/62344818/axon-improperly-handling-merged-segments

If you require further info please let me know.

```
2020-06-12 14:41:51.417  INFO 14056 [:] --- [EventHandler]-0] o.h.e.internal.DefaultLoadEventListener  : HHH000327: Error performing load command : org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.axonframework.eventhandling.tokenstore.jpa.TokenEntry#org.axonframework.eventhandling.tokenstore.jpa.TokenEntry$PK@9e1d11cf]

org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.axonframework.eventhandling.tokenstore.jpa.TokenEntry#org.axonframework.eventhandling.tokenstore.jpa.TokenEntry$PK@9e1d11cf]
    at org.hibernate.dialect.lock.PessimisticWriteSelectLockingStrategy.lock(PessimisticWriteSelectLockingStrategy.java:76)
    at org.hibernate.persister.entity.AbstractEntityPersister.lock(AbstractEntityPersister.java:1928)
    at org.hibernate.event.internal.AbstractLockUpgradeEventListener.upgradeLock(AbstractLockUpgradeEventListener.java:82)
    at org.hibernate.event.internal.DefaultLoadEventListener.loadFromSessionCache(DefaultLoadEventListener.java:569)
    at org.hibernate.event.internal.DefaultLoadEventListener.doLoad(DefaultLoadEventListener.java:444)
    at org.hibernate.event.internal.DefaultLoadEventListener.load(DefaultLoadEventListener.java:222)
    at org.hibernate.event.internal.DefaultLoadEventListener.lockAndLoad(DefaultLoadEventListener.java:406)
    at org.hibernate.event.internal.DefaultLoadEventListener.doOnLoad(DefaultLoadEventListener.java:127)
    at org.hibernate.event.internal.DefaultLoadEventListener.onLoad(DefaultLoadEventListener.java:92)
    at org.hibernate.internal.SessionImpl.fireLoad(SessionImpl.java:1256)
    at org.hibernate.internal.SessionImpl.access$1900(SessionImpl.java:207)
    at org.hibernate.internal.SessionImpl$IdentifierLoadAccessImpl.doLoad(SessionImpl.java:2866)
    at org.hibernate.internal.SessionImpl$IdentifierLoadAccessImpl.load(SessionImpl.java:2847)
    at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3482)
    at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3456)
    at jdk.internal.reflect.GeneratedMethodAccessor219.invoke(Unknown Source)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:308)
    at com.sun.proxy.$Proxy277.find(Unknown Source)
    at org.axonframework.eventhandling.tokenstore.jpa.JpaTokenStore.loadToken(JpaTokenStore.java:216)
    at org.axonframework.eventhandling.tokenstore.jpa.JpaTokenStore.storeToken(JpaTokenStore.java:111)
    at org.axonframework.eventhandling.TrackingEventProcessor$MergeSegmentInstruction.runSafe(TrackingEventProcessor.java:1385)
    at org.axonframework.eventhandling.TrackingEventProcessor$Instruction.lambda$null$0(TrackingEventProcessor.java:1139)
    at org.axonframework.common.transaction.TransactionManager.executeInTransaction(TransactionManager.java:47)
    at org.axonframework.eventhandling.TrackingEventProcessor$Instruction.lambda$run$1(TrackingEventProcessor.java:1139)
    at org.axonframework.common.ProcessUtils.executeWithRetry(ProcessUtils.java:33)
    at org.axonframework.eventhandling.TrackingEventProcessor$Instruction.run(TrackingEventProcessor.java:1139)
    at org.axonframework.eventhandling.TrackingEventProcessor.processInstructions(TrackingEventProcessor.java:332)
    at org.axonframework.eventhandling.TrackingEventProcessor.processingLoop(TrackingEventProcessor.java:297)
    at org.axonframework.eventhandling.TrackingEventProcessor$TrackingSegmentWorker.run(TrackingEventProcessor.java:1161)
    at org.axonframework.eventhandling.TrackingEventProcessor$WorkerLauncher.run(TrackingEventProcessor.java:1276)
    at java.base/java.lang.Thread.run(Thread.java:834)

2020-06-12 14:41:51.431 ERROR 14056 [:] --- [ault-executor-0] o.a.a.c.p.EventProcessorControlService   : Failed to merge segment [3] for processor [MyEventHandler]

java.util.concurrent.CompletionException: javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.axonframework.eventhandling.tokenstore.jpa.TokenEntry#org.axonframework.eventhandling.tokenstore.jpa.TokenEntry$PK@9e1d11cf]
    at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
    at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
    at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:632)
    at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
    at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
    at org.axonframework.eventhandling.TrackingEventProcessor$Instruction.run(TrackingEventProcessor.java:1143)
    at org.axonframework.eventhandling.TrackingEventProcessor.processInstructions(TrackingEventProcessor.java:332)
    at org.axonframework.eventhandling.TrackingEventProcessor.processingLoop(TrackingEventProcessor.java:297)
    at org.axonframework.eventhandling.TrackingEventProcessor$TrackingSegmentWorker.run(TrackingEventProcessor.java:1161)
    at org.axonframework.eventhandling.TrackingEventProcessor$WorkerLauncher.run(TrackingEventProcessor.java:1276)
    at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.axonframework.eventhandling.tokenstore.jpa.TokenEntry#org.axonframework.eventhandling.tokenstore.jpa.TokenEntry$PK@9e1d11cf]
    at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:226)
    at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:93)
    at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:200)
    at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3515)
    at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3456)
    at jdk.internal.reflect.GeneratedMethodAccessor219.invoke(Unknown Source)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:308)
    at com.sun.proxy.$Proxy277.find(Unknown Source)
    at org.axonframework.eventhandling.tokenstore.jpa.JpaTokenStore.loadToken(JpaTokenStore.java:216)
    at org.axonframework.eventhandling.tokenstore.jpa.JpaTokenStore.storeToken(JpaTokenStore.java:111)
    at org.axonframework.eventhandling.TrackingEventProcessor$MergeSegmentInstruction.runSafe(TrackingEventProcessor.java:1385)
    at org.axonframework.eventhandling.TrackingEventProcessor$Instruction.lambda$null$0(TrackingEventProcessor.java:1139)
    at org.axonframework.common.transaction.TransactionManager.executeInTransaction(TransactionManager.java:47)
    at org.axonframework.eventhandling.TrackingEventProcessor$Instruction.lambda$run$1(TrackingEventProcessor.java:1139)
    at org.axonframework.common.ProcessUtils.executeWithRetry(ProcessUtils.java:33)
    at org.axonframework.eventhandling.TrackingEventProcessor$Instruction.run(TrackingEventProcessor.java:1139)
    ... 5 common frames omitted
```

**Comment by ptomli** [comment]
Further testing seems to indicate that the pause/resume workaround isn't reliable either. I think there's some periodic sync system between the server and application instances, and if things happen too fast, or simply out of order with the sync system, then it breaks.

**Comment by saratry** [comment]
@ptomli, I've moved this issue into AxonFramework repository, as the exception that you described is related to the framework implementation.
https://github.com/AxonFramework/AxonFramework/issues/1451

Investigating this issue, I also found another problem related to merging feature.
I created a new issue here for that specific problem:
https://github.com/AxonIQ/axon-server-se/issues/136

So I'm going to close this one.
